codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -214,7 +232,99 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
             dstPort = header.readUnsignedShort();
         }
 
-        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort);
+        final List<HAProxyTLV> tlvs = readTlvs(header);
+
+        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort, tlvs);
+    }
+
+    private static List<HAProxyTLV> readTlvs(final ByteBuf header) {
+
+        // We need at least 4 byte for a valid TLV. If we have less bytes available, we need to ignore
+        if (header.readableBytes() < 4) {
+            return Collections.emptyList();
+        }
+
+        // In most cases there are less than 4 TLVs available
+        final List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>(4);
+
+        while (header.isReadable()) {
+            final HAProxyTLV haProxyTLV = readNextTLV(header);
+
+            if (haProxyTLV == null) {
+                // We don't have a TLV available anymore, we can stop
+                break;
+            }
+            haProxyTLVs.add(haProxyTLV);
+            if (haProxyTLV instanceof HAProxySSLTLV) {
+                haProxyTLVs.addAll(((HAProxySSLTLV) haProxyTLV).encapsulatedTLVs());
+            }
+        }
+
+        return haProxyTLVs;
+    }
+
+    private static HAProxyTLV readNextTLV(final ByteBuf header) {
+
+        // We need at least 4 bytes for a TLV
+        if (header.readableBytes() < 4) {
+            return null;
+        }
+
+        final byte typeAsByte = header.readByte();
+        final HAProxyTLV.Type type = HAProxyTLV.Type.typeForByteValue(typeAsByte);
+
+        final int length = header.readUnsignedShort();
+        switch (type) {
+        case PP2_TYPE_SSL:
+            final byte[] rawContent = copyByteBufToByteArray(header, length);
```

Comment:
@dobermai please extend `ByteBufHolder` ... If a user needs a byte[] its possible to do a memory copy later on.